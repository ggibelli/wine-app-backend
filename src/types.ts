export enum Province {
  AG = 'AG',
  AL = 'AL',
  AN = 'AN',
  AO = 'AO',
  AR = 'AR',
  AP = 'AP',
  AT = 'AT',
  AV = 'AV',
  BA = 'BA',
  BL = 'BL',
  BN = 'BN',
  BG = 'BG',
  BI = 'BI',
  BO = 'BO',
  BZ = 'BZ',
  BS = 'BS',
  BR = 'BR',
  CA = 'CA',
  CL = 'CL',
  CB = 'CB',
  CE = 'CE',
  CT = 'CT',
  CZ = 'CZ',
  CH = 'CH',
  CO = 'CO',
  CS = 'CS',
  CR = 'CR',
  KR = 'KR',
  CN = 'CN',
  EN = 'EN',
  FM = 'FM',
  FE = 'FE',
  FI = 'FI',
  FG = 'FG',
  FO = 'FO',
  FC = 'FC',
  FR = 'FR',
  GE = 'GE',
  GO = 'GO',
  GR = 'GR',
  IM = 'IM',
  IS = 'IS',
  SP = 'SP',
  AQ = 'AQ',
  LT = 'LT',
  LE = 'LE',
  LC = 'LC',
  LI = 'LI',
  LO = 'LO',
  LU = 'LU',
  MC = 'MC',
  MN = 'MN',
  MS = 'MS',
  MT = 'MT',
  ME = 'ME',
  MI = 'MI',
  MO = 'MO',
  MB = 'MB',
  NA = 'NA',
  NO = 'NO',
  NU = 'NU',
  OR = 'OR',
  PD = 'PD',
  PA = 'PA',
  PR = 'PR',
  PV = 'PV',
  PG = 'PG',
  PS = 'PS',
  PU = 'PU',
  PE = 'PE',
  PC = 'PC',
  PI = 'PI',
  PT = 'PT',
  PN = 'PN',
  PZ = 'PZ',
  PO = 'PO',
  RG = 'RG',
  RA = 'RA',
  RC = 'RC',
  RE = 'RE',
  RI = 'RI',
  RN = 'RN',
  RM = 'RM',
  RO = 'RO',
  SA = 'SA',
  SS = 'SS',
  SV = 'SV',
  SI = 'SI',
  SR = 'SR',
  SO = 'SO',
  TA = 'TA',
  TE = 'TE',
  TR = 'TR',
  TO = 'TO',
  TP = 'TP',
  TN = 'TN',
  TV = 'TV',
  TS = 'TS',
  UD = 'UD',
  VA = 'VA',
  VE = 'VE',
  VB = 'VB',
  VC = 'VC',
  VR = 'VR',
  VV = 'VV',
  VI = 'VI',
  VT = 'VT',
}

export enum Regioni {
  ABRUZZO = 'ABRUZZO',
  BASILICATA = 'BASILICATA',
  CALABRIA = 'CALABRIA',
  CAMPANIA = 'CAMPANIA',
  EMILIA = 'EMILIA',
  FRIULI = 'FRIULI',
  LAZIO = 'LAZIO',
  LIGURIA = 'LIGURIA',
  LOMBARDIA = 'LOMBARDIA',
  MARCHE = 'MARCHE',
  MOLISE = 'MOLISE',
  PIEMONTE = 'PIEMONTE',
  PUGLIA = 'PUGLIA',
  SARDEGNA = 'SARDEGNA',
  SICILIA = 'SICILIA',
  TOSCANA = 'TOSCANA',
  BOLZANO = 'BOLZANO',
  TRENTO = 'TRENTO',
  UMBRIA = 'UMBRIA',
  VALDAOSTA = 'VALDAOSTA',
  VENETO = 'VENETO',
}

export interface Address {
  via: string;
  CAP: number;
  comune: string;
  provincia: Province;
  regione: Regioni;
}

export enum TypeAd {
  SELL = 'SELL',
  BUY = 'BUY',
}

export enum MetodoProduttivo {
  CONVENZIONALE = 'CONVENZIONALE',
  BIOLOGICO = 'BIOLOGICO',
  BIODINAMICO = 'BIODINAMICO',
  NATURALE = 'NATURALE',
}

export enum TypeProduct {
  ADWINE = 'AdWine',
  ADGRAPE = 'AdGrape',
}

export enum EspressioneComunitaria {
  DOP = 'DOP',
  IGP = 'IGP',
  ND = 'ND',
}

export enum DenomZona {
  DOC = 'DOC',
  DOCG = 'DOCG',
  IGT = 'IGT',
  VARIETALE = 'VARIETALE',
  VINO = 'VINO',
}

export enum Menzione {
  CLASSICO = 'CLASSICO',
  RISERVA = 'RISERVA',
  SUPERIORE = 'SUPERIORE',
  VIGNA = 'VIGNA',
}

export enum QueryOrderBy {
  createdAt_ASC = 'createdAt_ASC',
  createdAt_DESC = 'createdAt_DESC',
  price_ASC = 'price_ASC',
  price_DESC = 'price_DESC',
}

export interface Errors {
  text: string;
  name: string;
}

export enum Rating {
  POOR = 'POOR',
  AVERAGE = 'AVERAGE',
  OK = 'OK',
  GOOD = 'GOOD',
  PERFECT = 'PERFECT',
}

export enum Colore {
  BIANCA = 'BIANCA',
  ROSSA = 'ROSSA',
}
