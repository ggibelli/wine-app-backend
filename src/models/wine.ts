import mongoose, { Schema, Document } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Regioni, EspressioneComunitaria, DenomZona } from '../types';
import { Regioni as MongooseRegioni } from '../utils/enumMongooseHelper';

export interface IWine {
  denominazioneVino: string;
  aka?: string;
  espressioneComunitaria: EspressioneComunitaria;
  denominazioneZona: DenomZona;
  regione: mongoose.Types.Array<Regioni>;
}

export interface WineGraphQl extends IWine {
  _id: mongoose.Types.ObjectId;
}

export interface IWineDoc extends IWine, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const wineSchemaFields: Record<keyof IWine, any> = {
  denominazioneVino: {
    type: String,
    required: true,
    minlength: 5,
    unique: true,
  },
  aka: String,
  espressioneComunitaria: {
    type: String,
    enum: ['DOP', 'IGP', 'ND'],
  },
  denominazioneZona: {
    type: String,
    enum: ['DOC', 'DOCG', 'IGT', 'Vino varietale', 'Vino generico'],
  },
  regione: [
    {
      type: String,
      enum: Object.values(MongooseRegioni),
      required: true,
    },
  ],
};

const wineSchema = new Schema(wineSchemaFields);

wineSchema.index({ denominazioneVino: 1 });

wineSchema.plugin(mongooseUniqueValidator);

export const Wine = mongoose.model<IWineDoc>('Wine', wineSchema);
