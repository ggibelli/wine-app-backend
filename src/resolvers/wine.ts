import { Resolvers, WineInput, WineInputUpdate } from '../generated/graphql';
import Users from '../data-sources/users';
import Wines from '../data-sources/wines';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  wines: Wines;
}

const resolver: StringIndexed<Resolvers> = {
  Query: {
    async wines(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.wines.getWines();
    },
    async wine(_, { id }, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.wines.getWine(id);
    },
  },
  Mutation: {
    async createWine(
      _,
      { wine }: { wine: WineInput },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.createWine(wine);
    },

    async updateWine(
      _,
      { wine }: { wine: WineInputUpdate },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.updateWine(wine);
    },
    async deleteWine(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.deleteWine(id);
    },
  },
};

export default resolver;
