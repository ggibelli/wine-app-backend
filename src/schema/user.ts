import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type ProducedWines {
    wine: Wine!
    bottlesProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  type OwnedVineyards {
    vineyard: Vineyard!
    tonsProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  input ProducedWinesInput {
    wine: ID!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input OwnedVineyardsInput {
    vineyard: ID!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input UserInput {
    email: String!
    password: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: AddressInput!
    hideContact: Boolean!
    #verified: Boolean
    #premium: Boolean
    #ads: [ID!]
    #negotiations: [ID!]
    #reviews: [ID!]
    #adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  input UserInputUpdate {
    _id: ID!
    email: String
    password: String
    firstName: String
    lastName: String
    pIva: String
    phoneNumber: String
    address: AddressInput
    isVerified: Boolean
    isPremium: Boolean
    hideContact: Boolean!
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  type Coordinates {
    latitude: Float
    longitude: Float
  }

  type User {
    _id: ID!
    email: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: Address!
    isVerified: Boolean!
    coordinates: Coordinates
    isPremium: Boolean
    isAdmin: Boolean!
    hideContact: Boolean!
    ads: [Ad!]
    savedAds: [Ad!]
    messages: [Message!]
    negotiations: [Negotiation!]
    reviews: [Review!]
    adsRemaining: Int @authorized
    dateCreated: Date! @date
    producedWines: ProducedWines
    ownedVineyards: OwnedVineyards
  }

  type AuthUser {
    token: String!
    user: User!
  }

  type AuthUserPayload {
    response: AuthUser
    errors: [Errors]
  }

  type UserPayload {
    response: User
    errors: [Errors]
  }

  extend type Query {
    users: [User!] @authenticated
    user(id: ID!): User @authenticated
    me: User @authenticated
  }

  extend type Mutation {
    createUser(user: UserInput!): AuthUserPayload
    updateUser(user: UserInputUpdate!): UserPayload @authenticated
    deleteUser(id: ID!): UserPayload @authenticated
    login(email: String!, password: String!): AuthUserPayload
  }
`;
