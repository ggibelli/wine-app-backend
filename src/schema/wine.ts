import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  input WineInput {
    denominazioneVino: String!
    aka: String
    espressioneComunitaria: EspressioneComunitaria!
    denominazioneZona: DenomZona!
    regione: [Regioni!]!
  }

  input WineInputUpdate {
    _id: ID!
    denominazioneVino: String
    aka: String
    espressioneComunitaria: EspressioneComunitaria
    denominazioneZona: DenomZona
    regione: [Regioni]
  }

  type Wine {
    _id: ID!
    denominazioneVino: String!
    aka: String
    espressioneComunitaria: EspressioneComunitaria!
    denominazioneZona: DenomZona!
    regione: [Regioni!]!
  }

  type WinePayload {
    response: Wine
    errors: [Errors]
  }

  extend type Query {
    wines: [Wine!] @authenticated
    wine(id: ID!): Wine @authenticated
  }

  extend type Mutation {
    createWine(wine: WineInput!): WinePayload @authorized @authenticated
    updateWine(wine: WineInputUpdate!): WinePayload @authorized @authenticated
    deleteWine(id: ID!): WinePayload @authorized @authenticated
  }
`;
