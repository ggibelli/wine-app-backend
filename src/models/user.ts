/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
