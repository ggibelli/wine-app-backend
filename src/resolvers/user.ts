import { Resolvers, UserInput, UserInputUpdate } from '../generated/graphql';
import { UserGraphQl } from '../models/user';
import Users from '../data-sources/users';
import Ads from '../data-sources/ads';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { ForbiddenError } from 'apollo-server-express';
import Reviews from '../data-sources/reviews';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  ads: Ads;
  messages: Messages;
  negotiations: Negotiations;
  reviews: Reviews;
  // wines: Wines;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async users(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.users.getUsers();
    },
    async user(_, { id }, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.users.getUser(id);
    },
    me(_, __, { user }: { user: UserGraphQl }) {
      return user;
    },
  },
  Mutation: {
    async createUser(
      _,
      { user }: { user: UserInput },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.createUser(user);
    },
    async updateUser(
      _,
      { user }: { user: UserInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.updateUser(user);
    },
    async deleteUser(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.deleteUser(id);
    },
    async login(
      _,
      { email, password }: { email: string; password: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      return dataSources.users.login(email, password);
    },
  },
  User: {
    async ads(user, _, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.ads.getAdsByUser(user._id);
    },
    async messages(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.messages.getMessages();
      }
      throw new ForbiddenError('You can only see your own messages');
    },
    async negotiations(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.negotiations.getNegotiations();
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    async reviews(
      root,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ) {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.reviews.getReviews();
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    email(root, _, { user }: { user: UserGraphQl }) {
      if (root._id.toString() === user._id.toString()) {
        return root.email;
      }
      if (!root.hideContact) {
        return root.email;
      }
      return 'The user prefers to hide the email';
    },
    phoneNumber(root, _, { user }: { user: UserGraphQl }) {
      if (root._id.toString() === user._id.toString()) {
        return root.phoneNumber;
      }
      if (!root.hideContact) {
        return root.phoneNumber;
      }
      return 'The user prefers to hide the phone number';
    },
    // aggiungo logica per mostrare solo negotiation e certi dati solo se sono i miei
  },
};
