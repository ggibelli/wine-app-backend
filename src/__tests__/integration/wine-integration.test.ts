/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
//import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { users } from '../../tests/mocksTests';
import { WineInput, WineInputUpdate } from '../../generated/graphql';
import { Regioni, EspressioneComunitaria, DenomZona } from '../../types';
// import createWineDb from '../../utils/wineExtractor';
import { Wine, WineGraphQl } from '../../models/wine';

const { query, mutate, setOptions } = testClient;

const WINES = gql`
  {
    wines {
      denominazioneVino
      espressioneComunitaria
      regione
      denominazioneZona
    }
  }
`;

const WINE = gql`
  query GetWine($id: ID!) {
    wine(id: $id) {
      denominazioneVino
      espressioneComunitaria
      regione
      denominazioneZona
    }
  }
`;

const CREATE_WINE = gql`
  mutation CreateWine($wine: WineInput!) {
    createWine(wine: $wine) {
      response {
        denominazioneVino
        espressioneComunitaria
        regione
        denominazioneZona
      }
      errors {
        name
        text
      }
    }
  }
`;

const UPDATE_WINE = gql`
  mutation UpdateWine($wine: WineInputUpdate!) {
    updateWine(wine: $wine) {
      response {
        denominazioneVino
        espressioneComunitaria
        regione
        denominazioneZona
      }
      errors {
        name
        text
      }
    }
  }
`;

const DELETE_WINE = gql`
  mutation DeleteWine($id: ID!) {
    deleteWine(id: $id) {
      response {
        denominazioneVino
        espressioneComunitaria
        regione
        denominazioneZona
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const user = new User(usersMock[0]);
  const otherUser = new User(usersMock[1]);
  const wine = new Wine({
    denominazioneVino: 'Abruzzo',
    espressioneComunitaria: 'DOP',
    denominazioneZona: 'DOC',
    regione: [Regioni.ABRUZZO],
  });
  await user.save();
  await otherUser.save();
  await wine.save();
});

describe('Integration test wines', () => {
  it('query wines fails if not logged in', async () => {
    const res = await query(WINES);
    expect(res).toMatchSnapshot();
  });

  it('query single wine fails if not logged', async () => {
    const wines: WineGraphQl[] = await Wine.find({}).lean().exec();
    const res = await query(WINE, {
      variables: { id: wines[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create wine mutation fails if not logged in', async () => {
    const wine: WineInput = {
      denominazioneVino: 'ciccio',
      espressioneComunitaria: EspressioneComunitaria.DOP,
      //@ts-ignore
      regione: [Regioni.CALABRIA],
      denominazioneZona: DenomZona.DOC,
    };
    const res = await mutate(CREATE_WINE, {
      variables: { wine },
    });
    expect(res).toMatchSnapshot();
  });

  it('query wines succeds if logged in', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(WINES);
    expect(res).toMatchSnapshot();
  });

  it('query single wine succeds if logged', async () => {
    const wine: WineGraphQl[] = await Wine.find({}).lean().exec();
    const res = await query(WINE, {
      variables: { id: wine[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create wine mutation successfull if logged in and admin', async () => {
    const wine: WineInput = {
      denominazioneVino: 'ciccio',
      espressioneComunitaria: EspressioneComunitaria.DOP,
      //@ts-ignore
      regione: [Regioni.CALABRIA],
      denominazioneZona: DenomZona.DOC,
    };
    const res = await mutate(CREATE_WINE, {
      variables: { wine },
    });
    expect(res).toMatchSnapshot();
  });

  it('update wine mutation succeds if logged in and admin', async () => {
    const wineToUpdate: WineGraphQl[] = await Wine.find({}).lean().exec();
    const wine: WineInputUpdate = {
      _id: wineToUpdate[0]._id.toString(),
      denominazioneVino: 'Caprotto',
    };
    const res = await mutate(UPDATE_WINE, {
      variables: { wine },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete wine mutation succeds if logged in and admin', async () => {
    const wineToDelete: WineGraphQl[] = await Wine.find({}).lean().exec();
    const res = await mutate(DELETE_WINE, {
      variables: { id: wineToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('update wine mutation fails if logged in and not admin', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const wineToUpdate: WineGraphQl[] = await Wine.find({}).lean().exec();
    const wine: WineInputUpdate = {
      _id: wineToUpdate[0]._id.toString(),
      denominazioneVino: 'Caprotto',
    };
    const res = await mutate(UPDATE_WINE, {
      variables: { wine },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete wine mutation fails if logged in and not admin', async () => {
    const wineToDelete: WineGraphQl[] = await Wine.find({}).lean().exec();
    const res = await mutate(DELETE_WINE, {
      variables: { id: wineToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create wine mutation fails if logged in and not admin', async () => {
    const wine: WineInput = {
      denominazioneVino: 'ciccio',
      espressioneComunitaria: EspressioneComunitaria.DOP,
      //@ts-ignore
      regione: [Regioni.CALABRIA],
      denominazioneZona: DenomZona.DOC,
    };
    const res = await mutate(CREATE_WINE, {
      variables: { wine },
    });
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
