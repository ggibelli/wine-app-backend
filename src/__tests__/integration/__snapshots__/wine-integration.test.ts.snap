// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test wines create wine mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "createWine": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "createWine",
      ],
    },
  ],
}
`;

exports[`Integration test wines create wine mutation fails if not logged in 1`] = `
Object {
  "data": Object {
    "createWine": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createWine",
      ],
    },
  ],
}
`;

exports[`Integration test wines create wine mutation successfull if logged in and admin 1`] = `
Object {
  "data": Object {
    "createWine": Object {
      "errors": Array [],
      "response": Object {
        "denominazioneVino": "ciccio",
        "denominazioneZona": "DOC",
        "espressioneComunitaria": "DOP",
        "regione": Array [
          "CALABRIA",
        ],
      },
    },
  },
}
`;

exports[`Integration test wines delete wine mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "deleteWine": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "deleteWine",
      ],
    },
  ],
}
`;

exports[`Integration test wines delete wine mutation succeds if logged in and admin 1`] = `
Object {
  "data": Object {
    "deleteWine": Object {
      "errors": Array [],
      "response": Object {
        "denominazioneVino": "Caprotto",
        "denominazioneZona": "DOC",
        "espressioneComunitaria": "DOP",
        "regione": Array [
          "ABRUZZO",
        ],
      },
    },
  },
}
`;

exports[`Integration test wines query single wine fails if not logged 1`] = `
Object {
  "data": Object {
    "wine": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "wine",
      ],
    },
  ],
}
`;

exports[`Integration test wines query single wine succeds if logged 1`] = `
Object {
  "data": Object {
    "wine": Object {
      "denominazioneVino": "Abruzzo",
      "denominazioneZona": "DOC",
      "espressioneComunitaria": "DOP",
      "regione": Array [
        "ABRUZZO",
      ],
    },
  },
}
`;

exports[`Integration test wines query wines fails if not logged in 1`] = `
Object {
  "data": Object {
    "wines": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "wines",
      ],
    },
  ],
}
`;

exports[`Integration test wines query wines succeds if logged in 1`] = `
Object {
  "data": Object {
    "wines": Array [
      Object {
        "denominazioneVino": "Abruzzo",
        "denominazioneZona": "DOC",
        "espressioneComunitaria": "DOP",
        "regione": Array [
          "ABRUZZO",
        ],
      },
    ],
  },
}
`;

exports[`Integration test wines update wine mutation fails if logged in and not admin 1`] = `
Object {
  "data": Object {
    "updateWine": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "FORBIDDEN",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to be Admin to continue",
      "path": Array [
        "updateWine",
      ],
    },
  ],
}
`;

exports[`Integration test wines update wine mutation succeds if logged in and admin 1`] = `
Object {
  "data": Object {
    "updateWine": Object {
      "errors": Array [],
      "response": Object {
        "denominazioneVino": "Caprotto",
        "denominazioneZona": "DOC",
        "espressioneComunitaria": "DOP",
        "regione": Array [
          "ABRUZZO",
        ],
      },
    },
  },
}
`;
